src/picongpu/include/fields/background/templates/bgrTWTS.hpp

```diff
@@ -0,0 +1,659 @@
+            if ( auto_tdelay ) {
+                
+                // angle between the laser pulse front and the y-axis. (Good approximation for beta0 \simeq 1. For exact relation look in TWTS core routines for Ex, By or Bz.)
+                const float_64 eta = PI/2 - (phi/2);
+                /* halfSimSize[2] --> Half-depth of simulation volume (in z); By geometric projection we calculate the y-distance walkoff
+                 * of the TWTS-pulse. The abs()-function is for correct offset for -phi<-90Deg and +phi>+90Deg. */
+                const float_64 y1=precisionCast<float_64>(halfSimSize[2]*picongpu::SI::CELL_DEPTH_SI)*abs(cos(eta)); 
+                const float_64 m=3.; // Fudge parameter to make sure, that TWTS pulse starts to impact simulation volume at low intensity values.
+                const float_64 y2=m*(pulselength_SI*picongpu::SI::SPEED_OF_LIGHT_SI)/cos(eta); // Approximate cross section of laser pulse through y-axis, scaled with "fudge factor" m.
+                const float_64 y3=focus_y_SI; // y-position of laser coordinate system origin within simulation.
+                const float_64 tdelay= (y1+y2+y3)/(picongpu::SI::SPEED_OF_LIGHT_SI*beta_0);
+                
+                return time_SI-tdelay;
+            }
+            else
+                return time_SI-tdelay_user_SI;
+        }
+        
+        template<>
+        HDINLINE float3_X
+        TWTSFieldE::getTWTSEfield_SI<DIM3>( const PMacc::math::Vector<float3_64,FieldE::numComponents>& eFieldPositions_SI, const float_64 time) const
+        {
+            return float3_X( float_X( calcTWTSEx(eFieldPositions_SI[0],time)/unitField[0] ), float_X(0.), float_X(0.) );
+        }
+        
+        template<>
+        HDINLINE float3_X
+        TWTSFieldE::getTWTSEfield_SI<DIM2>( const PMacc::math::Vector<float3_64,FieldE::numComponents>& eFieldPositions_SI, const float_64 time) const
+        {
+            // Ex->Ez, so also the grid cell offset for Ez has to be used.
+            return float3_X( float_X(0.), float_X(0.), precisionCast<float_X>( calcTWTSEx(eFieldPositions_SI[2],time)/unitField[2] ) );
+        }
+        
+        HDINLINE float3_X
+        TWTSFieldE::operator()( const DataSpace<simDim>& cellIdx,
+                                const uint32_t currentStep ) const
+        {
+            const float_64 time_SI=getTime_SI(currentStep);
+            const PMacc::math::Vector<float3_64,FieldE::numComponents> eFieldPositions_SI=getEfieldPositions_SI<simDim>(cellIdx);
+            // Single TWTS-Pulse
+            return getTWTSEfield_SI<simDim>(eFieldPositions_SI, time_SI);
+        }
+
+        /** Calculate the Ex(r,t) field here
+         *
+         * \param pos Spatial position of the target field.
+         * \param time Absolute time (SI, including all offsets and transformations) for calculating the field */
+        HDINLINE float_64
+        TWTSFieldE::calcTWTSEx( const float3_64& pos, const float_64 time) const
```

Comment:
a general remark: if you calculate the factors below here all accordingly, your calculation can be speeded up by a factor 3 since you can in principle use `float_X` for your operations.